terraform/state.go

```diff
@@ -1876,11 +1877,19 @@ var ErrNoState = errors.New("no state")
 // ReadState reads a state structure out of a reader in the format that
 // was written by WriteState.
 func ReadState(src io.Reader) (*State, error) {
+	// check for a nil file specifically, since that produces a platform
+	// specific error if we try to use it in a bufio.Reader.
+	if f, ok := src.(*os.File); ok && f == nil {
+		return nil, ErrNoState
+	}
+
 	buf := bufio.NewReader(src)
+
 	if _, err := buf.Peek(1); err != nil {
-		// the error is either io.EOF or "invalid argument", and both are from
-		// an empty state.
-		return nil, ErrNoState
+		if err == io.EOF {
```

Comment:
The removed comment here talks about both `io.EOF` and "invalid argument". Is the "invalid argument" case taken care of by the nil check above, or taken care of some other way here?

Reply:
oh yes, "invalid argument" is only expected when we have a nil `*os.File`. At one point that was ok to use for `ReadState` to indicate the absence of state. I'd like to not have that at all, but that takes some more serious auditing to make sure we don't have any more code paths passing it in.